Adds Taproot BIP341 signature message and create a unified sighash cache for legacy, segwit and taproot inputs

RCasatta opened the pull request:
Adds https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki message signature algorithm

The base is taken from `bip143::SigHashCache`, some code results duplicated but I think it's more clear to keep things separated

Would mark some bullet point on https://github.com/rust-bitcoin/rust-bitcoin/issues/503

Test vectors are taken by running https://github.com/bitcoin/bitcoin/blob/d1e4c56309aeb73772e3a9d779a9c157024c9e1e/test/functional/feature_taproot.py with a modified `TaprootSignatureHash` function to print intermediate values that I cannot found in the bip341 [test vector json](https://raw.githubusercontent.com/bitcoin-core/qa-assets/main/unit_test_data/script_assets_test.json)

UPDATE: Latest version includes the suggestion from @sanket1729 to create a unified tool for signature message hash for legacy, segwit, and taproot inputs. In particular, makes sense for mixed segwit v0 and taproot v1 inputs because cached values could be shared

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+
+/// Contains outputs of previous transactions to provide to the [SigHashCache::signature_hash]
+/// method, In the case [SigHashType] variant is `ANYONECANPAY`, [Prevouts::Anyone] may be provided
+pub enum Prevouts {
```
I think this should take in a reference and have a lifetime. 
```
pub enum Prevouts<'spent_uxtos> {
	Anyone(.. &'spent_utxo TxOut),
	All (&'spent_utxos [TxOut])
}
```

elichai replied:
Or maybe a Cow :)

RCasatta replied:
Is there any reason to prefer `Cow` over borrow since there are no clones/take ownership inside?

RCasatta replied:
added lifetime in 4da0c07

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+/// Information related to the script path spending
+pub struct ScriptPath {
+    script: Script,
+    code_separator: i32,
```
nit: slight preference to using `u32` because it is consistent with BIP342 description and c++ codebase.

sanket1729 replied:
Preference for `code_separator_pos` over `code_separtor`.

RCasatta replied:
> nit: slight preference to using `u32` because it is consistent with BIP342 description and c++ codebase.

in the python test code the default value is `-1` that's why I used i32.

are you suggesting to change to `u32` and use a default value of `0xFFFFFFFF` ?

sanket1729 replied:
Yeah, from BIP342

```
codesep_pos (4): the opcode position of the last executed OP_CODESEPARATOR before the currently 
executed signature opcode, with the value in little-endian (or 0xffffffff if none executed)
```

RCasatta replied:
renamed and changed type in e9d3777

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+
+/// Information related to the script path spending
+pub struct ScriptPath {
+    script: Script,
```
I also think this should be take in a script reference `&'script Script`

RCasatta replied:
done in 3d9422d

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        annex: Option<Vec<u8>>,
```
Annex should be taken in as reference

sanket1729 replied:
It may be overkill for now, but we can an Annex type which actually enforces that the first byte is `0x50`.

RCasatta replied:
I tried to add an "Annex type MVP" in 8098975

It should later be improved but it's nice if it's already a dedicated struct in the method signature

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+        sighash_type: SigHashType,
+    ) -> Result<(), Error> {
+        if let Prevouts::All(prevouts) = prevouts {
+            if prevouts.len() != self.tx.input.len() {
```
We should also check `input_index < tx.input.len()`

RCasatta replied:
That one should be covered by the `Prevouts::get` method, isn't it?

sanket1729 replied:
I am suggesting a check for `let txin = &self.tx.input[input_index];` not for prevouts

RCasatta replied:
added a new Error variant and the check in 00c5cf3

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+/// method, In the case [SigHashType] variant is `ANYONECANPAY`, [Prevouts::Anyone] may be provided
+pub enum Prevouts {
+    /// When modifier `ANYONECANPAY` is provided, only prevout of the current input is needed
+    Anyone(usize, TxOut),
```
nit: Add a doccomment about `usize` being the current index of input

RCasatta replied:
done in 5e4b64c

JeremyRubin replied:
would it make sense to represent as a u32 as that matches Outpoint better?

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+
+        // If hash_type & 3 does not equal SIGHASH_NONE or SIGHASH_SINGLE:
+        //     sha_outputs (32): the SHA256 of the serialization of all outputs in CTxOut format.
+        if let SigHashType::All | SigHashType::AllPlusAnyoneCanPay = sighash {
```
The code as is will fail after we add `SigHashType::Default`.

RCasatta replied:
what do you suggest?

sanket1729 replied:
This is just a note, that we will need to update the implementation. 

If we are going to merge this before we have `SighashType::Default`, we can follow the spec

`If hash_type & 3 does not equal SIGHASH_NONE or SIGHASH_SINGLE:`


If we are going to merge this after we have `SighashType::Default`, then we can do it as 
```
SigHashType::All | SigHashType::AllPlusAnyoneCanPay | SigHashType::Default
```

We can do the first suggestion for now.

RCasatta replied:
done in 2c4f52c

sanket1729 reviewed (commented):
Is it worth having a single sighash cache for all output types. It is a breaking change, but it appears to be cleaner and could also be more efficient when we are spending segwitv0 and taproot outputs together(as implemented in bitcoin core).

```
let cache = SighashCache::new(&tx);
let legacy_hash = cache.sighash_legacy();
let segwitv0_sighash = cache.sighash_segwitv0(amt, script..);
let taphash = cache.sighash_taproot(&prevouts);
```

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+        prevouts: &Prevouts,
+        annex: Option<Vec<u8>>,
+        script_path: Option<ScriptPath>,
+        sighash_type: SigHashType,
```
From
https://github.com/rust-bitcoin/rust-bitcoin/pull/628/commits/a783ceab229174b39a7b5d6cea8550558309c78d#r667735143

Should we have input as `u8` for `SigHashType`? Part of me thinks that extending SigHashType with another enum for `SigHashType::Default` might be super breaking downstream change. 

I am not super thrilled about a new `SighashTypeTaproot`, but that might be best way forward.

RCasatta replied:
I think a new `SighashTypeTaproot` with the same variant as `SigHashType` plus `SighashTypeTaproot::Default` is the best way, yes.

Maybe the name is not crystal-clear since it is introduced with taproot but will be usable also for pre-taproot sighashes

sanket1729 replied:
Yup agreed, +1 for `SighashTypeTaproot`(or any other better name)

JeremyRubin replied:
it might make sense to just enum it because sighash types can diverge in the future (e.g., SighashTypeAnyprevout

RCasatta commented:
> but it appears to be cleaner and could also be more efficient when we are spending segwitv0 and taproot outputs together

it is surely cleaner especially because it is handier to use (in case of mixed inputs it allows to instantiate only one sighash cache) but in what ways is it more efficient? The hashes are differents (both in order and also single instead of double)

sanket1729 commented:
That being said, I am also okay if we decide that the code complexity of this particular optimization is not worth the benefit. 

`rust-bitcoin` is not a full node that needs to quickly verify the transaction, so keeping things simple and easy to review might have its pros. My vote is towards implementing it, but I am also okay if others don't see the value in it.

sanket1729 reviewed (commented):
Thanks for doing this :) 
Code review ACK 8098975f301cfb6a08321d6183b3b26a0c809186. 

1) My vote is for adding a new SigHashType for Taproot. Initially, it seemed a bit wasteful at first to have a new SigHashType for every new sighash update(we will have another enum for ANYPREVOUT). But I am convinced that this is the correct way to proceed. 
2) And a single cache for all signature hashes(legacy, segwitv0, and taproot) with a new name and optionally mark the current `bip143::SighashCache` as deprecated.

RCasatta commented:
> Thanks for doing this :)

Thanks for the review!

> 1. My vote is for adding a new SigHashType for Taproot. Initially, it seemed a bit wasteful at first to have a new SigHashType for every new sighash update(we will have another enum for ANYPREVOUT). But I am convinced that this is the correct way to proceed.

Still TODO

> 2. And a single cache for all signature hashes(legacy, segwitv0, and taproot) with a new name and optionally mark the current `bip143::SighashCache` as deprecated.

apart from deprecation, latest commits implement the unified cache with legacy, segwitv0 and taproot

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,1241 @@
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        sighash_type: U,
```
Wanted to change to SigHashType but it's incompatible with `let sighash_arr = endian::u32_to_array_le(sighash_type);` inside `Transaction::encode_signing_data_to`

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,1241 @@
+        &self,
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
```
Not sure about this, I thought it could be handy with transaction with mixed inputs you probably already have the Prevouts struct, so it's handy to have it here, on the other hand the value inside `TxOut` is not used

RCasatta replied:
done in 9f51591

RCasatta commented:
The latest changes should include all feedback received (deprecation of `bip143::SigHashCache` and new `SigHashType` with `Default` variant.

I am going to wait for @sanket1729 feedback round, if mostly ok, I think I am going to squash the commits for easier review and remove draft state (probably 3 commits: add encodable to sha256::Hash, add sighash, deprecate bip143::sighash)

sanket1729 commented on src/util/sighash.rs:
```diff
@@ -172,7 +197,49 @@ impl<'s> ScriptPath<'s> {
+impl From<SigHashType> for LegacySigHashType {
+    fn from(s: SigHashType) -> Self {
+        match s {
+            SigHashType::Default => LegacySigHashType::All,
```
I don't think we should be doing this. We are changing `0x00` to `0x01`.

RCasatta replied:
I removed the from implementation entirely in 993dae0, using LegacySigHashType for the segwits methods doesn't require it anymore

sanket1729 commented on src/util/sighash.rs:
```diff
@@ -172,7 +197,49 @@ impl<'s> ScriptPath<'s> {
+    /// Break the sighash flag into the "real" sighash flag and the ANYONECANPAY boolean
+    pub(crate) fn split_anyonecanpay_flag(self) -> (SigHashType, bool) {
+        match self {
+            SigHashType::Default => (SigHashType::Default, false),
```
We should fail when the user provides `Default` for segwitv0 sighash

RCasatta replied:
this is for the taproot sighash, is it correct?

sanket1729 replied:
Yeah sorry. I don't know why this comment is placed for this code block. But I wanted is fixed in 993dae01b35038cc49e216afff9525d1b870f419

sanket1729 commented on src/util/bip341.rs:
```diff
@@ -467,6 +459,14 @@ mod tests {
+    fn sizeof() {
+        assert_eq!(size_of::<Option<CommonCache>>(), 97);
+        assert_eq!(size_of::<Option<TaprootCache>>(), 65);
+        assert_eq!(size_of::<&Transaction>(), 8);
```
According to Andrew Poelstra, this is not the same across all machines. It is the same size as `usize`.

RCasatta replied:
Andrew is totally right!

I used the test for some reasoning on stack space usage, I think I can remove it.
done in a37de1a

JeremyRubin commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+pub enum Prevouts {
+    /// When modifier `ANYONECANPAY` is provided, only prevout of the current input is needed
+    Anyone(usize, TxOut),
+    /// When `ANYONECANPAY` is not provided, or the caller is handy giving all prevouts so he can reuse
```
comment gender ;)

RCasatta replied:
fixed in c232442

JeremyRubin commented on src/util/bip341.rs:
```diff
@@ -0,0 +1,556 @@
+    fn get(&self, input_index: usize) -> Result<&TxOut, Error> {
+        match self {
+            Prevouts::Anyone(index, prevout) => {
+                if input_index == *index {
```
BIP-118 doesn;'t commit to the input_index, so this might not be robust?

Maybe make get take an Option index and only check if it's provided

JeremyRubin commented on src/util/bip341.rs:
```diff
@@ -24,6 +24,29 @@ use std::{error, fmt, io};
+    common_cache: Option<CommonCache>,
+
+    /// Cache for taproot v1 inputs
+    taproot_cache: Option<TaprootCache>,
```
I think we might as well just always cache everything?

RCasatta replied:
As bitcoin core is doing it's more efficient to separate the cache this way so you can compute only what you need and reuse it according to the type of the inputs.

* for legacy: no cache is computed
* for taproot only: no segwit_cache is computed
* for segwit only: no taproot_cache is computed

Even if we don't have the same efficiency requirement as bitcoin core I think it's nice to have (and also @sanket1729 have a preference for that https://github.com/rust-bitcoin/rust-bitcoin/pull/628#issuecomment-878460252)

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -325,17 +338,18 @@ impl<'a> SigHashCache<'a> {
         input_index: usize,
-        script_code: &Script,
-        value: u64,
+        prevouts: &Prevouts,
```
👍  -- note that we could also provide a (version with) a codesep argument if we don't want to have to clone the script, or provide a Script Slice (equivalent).

RCasatta replied:
kept segwit method in line with existent in e76e5e1

@JeremyRubin it is already a borrowed Script?

JeremyRubin replied:
codesep in segwit requires a script slice -- who owns the script presently? If you don't have a codesep argument OR a script slice (and compute sep ext), then somewhere you're allocating extra.

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -25,11 +25,14 @@ use std::{error, fmt, io};
-    /// Access to transaction required for various introspection
-    tx: &'a Transaction,
+/// Efficiently calculates signature hash message for legacy, segwit and taproot inputs.
+pub struct SigHashCache<T: Deref<Target=Transaction>> {
```
Don't you need DerefMut for that?

JeremyRubin replied:
Ah I see, it's optional functionality...

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -87,6 +89,29 @@ pub struct ScriptPath<'s> {
+    /// to follow it.)
+    Single = 0x03,
+    /// 0x81: Sign all outputs but only this input
+    AllPlusAnyoneCanPay = 0x81,
```
Plus is a bit of a weird word since it's a logical AND of the functionality (bitwise OR is plus though).

sanket1729 replied:
For better or for worse, this is also the current API in the already existing SigHashType. IMO it's better to be consistent.

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -87,6 +89,29 @@ pub struct ScriptPath<'s> {
+/// Hashtype of an input's signature, encoded in the last byte of the signature
+/// Fixed values so they can be casted as integer types for encoding
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum SigHashType {
```
I think this should be a 2 level enum where we have
```
enum SigHashType {
    Legacy,
    SegwitV0,
    SegwitV1,
}
```
to make clear contexts where modes do not have/have different meanings.

sanket1729 replied:
I think we should just have different types of `SigHashTypeLegacy`, `SighashTypeSegwitv0`, `SigHashTypeTaproot. 

In what use case would you need a level 2 enum?

JeremyRubin replied:
I think that's OK -- I mean more that there should not be a "SigHashType" general enum that From coerces into other SigHashType* things. Espeically once we get into BIP-118, it would switch into a TryFrom because certain SigHashTypes would not be proper.

RCasatta replied:
> I think we should just have different types of SigHashTypeLegacy, SighashTypeSegwitv0, SigHashTypeTaproot.

why different types for legacy and segwit?

Is it a very bad idea to add a `RESERVED` variant to the `sighash::SigHashType` introduced in this PR to accommodate SIGHASH_PREVOUT when it's time?
`#[non_exhaustive]` would be better but it's not in our MSRV

sanket1729 replied:
Sorry, I think I misread your comment initially for something else. I am not against adding the reserved variant, but I don't know if how effectively it serves the same purpose as `#[non-exhaustive]`. 

I will let other contributors comment on this, I am not sure if we should so much forward-thinking about `OP_ANYPREVOUT` right now. The current spec for BIP118 is itself a major change in PublicKey API because it introduces new key types. Maybe additionally doing this is that not that bad?

JeremyRubin replied:
I guess if I can better specify my position, I think that that one should be able to make a "pure" function:

```rust
fn get_sighash(tx:&TransactionData, sighash: SigHashType) -> Hash
```

which computes the correct digest with no additional instructions. Since legacy, segwitv0, and segwitv1 all compute different things, unifying the type as one `SigHashType` means that we need additional context, so

```rust
fn get_sighash_legacy(tx:&TransactionData, sighash: SigHashType) -> Hash
fn get_sighash_v0(tx:&TransactionData, sighash: SigHashType) -> Hash
fn get_sighash_v1(tx:&TransactionData, sighash: SigHashType) -> Hash
``` 

This is already sort of the case because v1 requires more context to sign, but I imagine the APIs will converge on having the superset of data available for any case (v1/v0/legacy).  I think you in theory could parse what digest version is required from the TxData, but it'd be nice if SigHashType were an explicit & unique digest instruction.

JeremyRubin replied:
I do agree this isn't a huge issue for now, just sounding the warning that BIP-118's plans do include diverging sighash types where there will be different things available at different times, and I think it is something we can handle relatively elegantly by being more explicit here.

dr-orlovsky replied:
`non_exhaustive` does not work on current MSRV

RCasatta commented on src/util/bip341.rs:
```diff
@@ -304,67 +310,50 @@ impl<'a> SigHashCache<'a> {
+            };
+            self.common_cache = Some(cache);
+        }
+        self.common_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
```
I changed it for the following reasons:

* the `segwit_cache()` method is now calling the `common_cache()` method that takes `&self`, so it's not possible anymore to borrow and mutable borrow at the same time (before was possible because we were borrowing and mutable borrowing on different struct 34fields).
* `common_cache()` and `taproot_cache()` could indeed use  `get_or_insert_with()` but it requires let bindings before the call that are done even in the cache hit phase, so I think it's a little better to avoid those (unless someone tells me they are optimized away by the compiler).
* since  it seems `segwit_cache()` cannot be implemented with `get_or_insert_with()`, I thought to be coherent with other methods

Unrelated to the `get_or_insert_with()` discussion, the method before was returning owned values, copying 32 bytes every time, while now it returns only a borrow

Kixunil replied:
I really don't believe there's performance loss. It is also possible to use `get_or_insert_with` in `segwit_cache` with some refactoring - see last commit of https://github.com/RCasatta/rust-bitcoin/pull/3

RCasatta commented on src/util/sighash.rs:
```diff
@@ -622,4 +721,125 @@ mod tests {
+
+    fn p2pkh_hex(pk: &str) -> Script {
+        let pk = Vec::from_hex(pk).unwrap();
+        let pk = PublicKey::from_slice(pk.as_slice()).unwrap();
```
done in a810722

RCasatta commented on src/util/sighash.rs:
```diff
@@ -385,12 +399,11 @@ impl<'a> SigHashCache<'a> {
     ) -> SigHash {
         let mut enc = SigHash::engine();
-        self.segwit_encode_signing_data_to(&mut enc, input_index, script_code, value, sighash_type)
+        self.segwit_encode_signing_data_to(&mut enc, input_index, prevouts, sighash_type)
```
Restored the previous API in e76e5e1

> We are now allowing this to panic when the prevouts struct is incorrectly specified

do you mean throwing errors? The intention is that methods in the sighash module never panic

sanket1729 replied:
Restoring to the previous API resolves my concerns.

RCasatta commented on src/util/sighash.rs:
```diff
@@ -318,7 +318,7 @@ impl<'a> SigHashCache<'a> {
 
     /// Compute the BIP341 sighash for any flag type.
-    pub fn taproot_signature_hash(
+    pub fn taproot_sig_hash(
```
resolved has seen in https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r669820790

sanket1729 reviewed (approved):
code review ACK a37de1ade475e0c31c932121abaa7aec701b9987. 

I will review the tests carefully later, but I am super happy with the current API design. We can squash this and open it up for feedback from other contributors.

sanket1729 reviewed (commented):
How about changing `segwit_signature_hash()` to `segwitv0_signature_hash()`?

RCasatta commented:
I squashed the commits, in comparison to the pre-squash situation I left legacy and tests in their place with a trick for legacy tests to tests also the new API code paths (while the segwit tests are left there since the `bip143::SigHashCache` impl is provided by `sighash::SigHashCache`)

I also added the proposed `RESERVED` variant in the last commit since I am not super sure about it

> How about changing `segwit_signature_hash()` to `segwitv0_signature_hash()`?

I am not super in love with `segwitv0` (it should be `taprootv1` for the other? But let's see what others think about it

sanket1729 reviewed (approved):
tACK 877dc473070301e8b0ce6f88413a9ab3f688a69f. Ran tests, checked that tests have covered all paths of interest, all sighash types, annex, scriptpaths. 

The only thing I am unsure about is the effectiveness of the `Reserved` variant.

sgeisler commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+        //      includes the mandatory 0x50 prefix.
+        if let Some(annex) = annex {
+            let mut enc = sha256::Hash::engine();
+            annex.as_bytes().to_vec().consensus_encode(&mut enc)?;
```
This allocation is unnecessary, I'd propose implementing `Encodable` for `Annex` instead.

RCasatta replied:
`impl Encodable for Annex ` added in 16a236b

sgeisler commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+            let hash = TapLeafHash::from_engine(enc);
+
+            hash.into_inner().consensus_encode(&mut writer)?;
+            0u8.consensus_encode(&mut writer)?;
```
Maybe make `KEY_VERSION_0` a constant or add a comment? It took me a while to find what its purpose is.

RCasatta replied:
added in 95fd41c

sgeisler reviewed (approved):
I left some nits, but generally utACK 877dc47

sanket1729 commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+    }
+
+    fn common_cache(&mut self) -> &CommonCache {
+        if self.common_cache.is_none() {
+            let mut enc_prevouts = sha256::Hash::engine();
+            let mut enc_sequences = sha256::Hash::engine();
+            for txin in self.tx.input.iter() {
+                txin.previous_output
+                    .consensus_encode(&mut enc_prevouts)
+                    .unwrap();
+                txin.sequence.consensus_encode(&mut enc_sequences).unwrap();
+            }
+            let cache = CommonCache {
+                prevouts: sha256::Hash::from_engine(enc_prevouts),
+                sequences: sha256::Hash::from_engine(enc_sequences),
+                outputs: {
+                    let mut enc = sha256::Hash::engine();
+                    for txout in self.tx.output.iter() {
+                        txout.consensus_encode(&mut enc).unwrap();
+                    }
+                    sha256::Hash::from_engine(enc)
+                },
+            };
+            self.common_cache = Some(cache);
+        }
+        self.common_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
```
See https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r670262033

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+    /// Should never happen since we are always encoding, thus we are avoiding wrap the IO error
+    IoError,
```
I guess because `?` is more reassuring to see than `unwrap()` ?

Apart from jokes, I based the initial code on `bip143::SigHashCache::encode_signing_data_to` where it's used `?` when calling consensus_encode, but then is simply unwrapped in `bip143::SigHashCache::signature_hash` so it's equivalent to call unwrap any time and remove the `IoError` variant as I did in 9930df36a3674b1537f49b1be9392e96bd464462

sgeisler replied:
Yeah, I think since it would be a logic error on our side it's best to panic early. Upstream shouldn't have to bother with something that should never happen. Thx for fixing :)

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+}
+
+/// Possible errors in computing the signature message
+#[derive(Debug, Eq, PartialEq)]
```
Derived common traits in 6e053b724c68515ab21b9a97def9c18a2b88202a

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+    IoError,
+
+    /// Requested input index is greater than the number of inputs in the given transaction
+    IndexGreaterThanInputsSize,
```
done in c3fd729eca75b6b17731bf180bec89cb15028f93

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+        }
+    }
+
+    fn check_index(&self, index: usize) -> Result<(), Error> {
```
Implementing this helped me find a validation bug in the case of SIGHASH_SINGLE when there is no corresponding output, according to the bip341 spec it's a validation error. 
Instead, it was allowed (and bound checked) in segwit by looking at the code.

Note that I left a check in the legacy case because I don't want to change the signature of the method in [Transaction::encode_signing_data_to] but I want panic-free code here

done in 4c1aaa6

sgeisler reviewed (dismissed):
utACK 6d4db61

sanket1729 reviewed (approved):
reACK 6d4db6112a49855034e51703e8267f4c51904cef

stevenroose commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,912 @@
+        let (sighash, anyone_can_pay) = sighash_type.split_anyonecanpay_flag();
+
+        // epoch
+        0u8.consensus_encode(&mut writer).unwrap();
```
This is assuming the `writer` is an `io::Write` that doesn't error, right? I suppose that should at least be documented, but ideally we'd propagate that error as an `Error::Writer(io::Error)` variant.

stevenroose replied:
I realize that we use this method internally with a safe Write, but since it's a public method, someone else might not..

RCasatta replied:
There was a variant, but there was a suggestion to remove it https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r671166566 since probably never exercised. Now I see calling `*_encode_signing*` with custom Writer may throw an Error, @sgeisler I think I need to remove 6d4db61 and change the variant comment, do you agree?

sgeisler replied:
Right, I totally missed these methods were public. So either remove d4db61 or make them private I guess (was is their intended public use case?).

stevenroose replied:
Good point that I think they might be really hard to use for anything else than hashing. The signing data is not very structured, so I doubt anyone wants to write it into anything else than a hasher. Maybe someone would want to use them with a hardware-accelerated sha256 engine or so?

RCasatta replied:
external writers are used for accessing serialized data, see https://github.com/rust-bitcoin/rust-bitcoin/issues/483

I removed 6d4db61 and added 1da9706db98b7e873b1209052cdb9648567dab7a

I wrapped `ErrorKind` instead of `io::Error` because I think it's a good tradeoff allowing to auto derive some useful traits (Copy, Eq...)

stevenroose replied:
ah there you go

stevenroose replied:
and yeah ack, io::Error is an annoying one.

sgeisler replied:
So this will probably require changes to #603 or the other way around depending on merge order.

stevenroose commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,912 @@
+            leaf_version,
+            code_separator_pos,
+        }) = script_path
+        {
```
nit, this fits in a single 93-character line which is shorter than the 100 char column limit, no?

RCasatta replied:
I just used default rustfmt for this file

sgeisler replied:
Oh wow, I wouldn't have complained here, but thought about leaving a nit for L345-353. Sometimes `rustfmt` seems quite off. But I think it's ok if it follows a consistent standard.

RCasatta replied:
since it needs re-approval, would you like 844dfe2dfce31e966daff9150ac613b734cc1f97 to improve fmt around line L345-353 ?

sgeisler replied:
Nah, we want to use `rustfmt` eventually anyway. If that's its output I'm fine with it.

RCasatta replied:
844dfe2 is always following `rustfmt` rules, but the added let binding change its behaviors

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,912 @@
+        Ok(SigHash::from_engine(enc))
+    }
+
+    fn common_cache(&mut self) -> &CommonCache {
```
I think it's personal taste unless there is some convention? I prefer to have principal method first

stevenroose replied:
In order languages this was a requirement, so maybe that's where a convention would have come from. I think it generally makes sense to aide a vertical reader. But again, personal preference.

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,912 @@
+
+impl error::Error for Error {}
+
+impl<'u> Prevouts<'u> {
```
I think it's personal taste unless there is some convention? I prefer to have structs and relations before impls

stevenroose replied:
Don't think there's a convention. Not sure if you use an IDE that has an outline sidebar or something? Without it, it's a bit annoying to have to go look through the file for the impls of a type. But yeah it's personal preference.

sgeisler commented on src/util/sighash.rs:
```diff
@@ -164,7 +165,7 @@ pub enum Error {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Error::IoError => write!(f, "IoError"),
+            Error::Io(ref e) => write!(f, "{:?}", e),
```
nit: context could be given here, ideally some message that can be grepped when a panic is thrown up-stack. This makes tracing back the root cause of errors much easier.

RCasatta replied:
like this? `write!(f, "sighash::Error::Io {:?}", e)`

sgeisler replied:
For example, or just a more verbal description like the other variants.

RCasatta replied:
using `write!(f, "Writer errored: {:?}", e)` amended last commit to 7b60763075a974ddd4812000ec90be4e17d3fb12

sgeisler commented on src/util/sighash.rs:
```diff
@@ -126,8 +126,9 @@ pub enum SigHashType {
-    IoError,
+    /// Could happen only by using `*_encode_signing_*` methods with custom writers, engines writers
+    /// like the ones used in methods `*_signature_hash` don't error
+    Io(io::ErrorKind),
```
Is there any particular reason for only keeping the kind?

RCasatta replied:
Only this
 https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r672234789

sgeisler replied:
Oh right, totally missed that, good point!

sgeisler reviewed (dismissed):
utACK 7b60763

sanket1729 reviewed (dismissed):
tACK 7b60763075a974ddd4812000ec90be4e17d3fb12

RCasatta commented:
Rebased because https://github.com/rust-bitcoin/rust-bitcoin/pull/603 has been merged

sanket1729 reviewed (approved):
reACK 6e06a32ccc70f13dc57c611cfcb6678ab520b9a0

sgeisler commented on src/util/bip143.rs:
```diff
@@ -107,121 +108,34 @@ impl SighashComponents {
     }
 
     /// Encode the BIP143 signing data for any flag type into a given object implementing a
     /// std::io::Write trait.
```
minor nit (please don't change): after #603 this is not entirely correct anymore

RCasatta replied:
fixed in c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f (Since other changes were needed)

sgeisler reviewed (approved):
reACK 6e06a32ccc70f13dc57c611cfcb6678ab520b9a0

dr-orlovsky commented on src/util/bip143.rs:
```diff
@@ -263,13 +178,14 @@ impl<R: DerefMut<Target=Transaction>> SigHashCache<R> {
 #[cfg(test)]
 #[allow(deprecated)]
 mod tests {
+    use std::str::FromStr;
```
I believe it should be
```suggestion
    use core::str::FromStr;
```

Kixunil replied:
AFAIK `std` in tests is OK.

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+
+/// Efficiently calculates signature hash message for legacy, segwit and taproot inputs.
+#[derive(Debug)]
+pub struct SigHashCache<T: Deref<Target = Transaction>> {
```
Not sure I understand why this does not use `AsRef<Transaction>` instead of `Deref<Target = Transaction>`.

Kixunil replied:
Or `Borrow` which would also allow for `SigHashCache<Transaction>` - not sure how important it is.

RCasatta replied:
We can't use `AsRef` because we need also a mutable borrow for `access_witness` and `AsMut` is not a subtrait of `AsRef` (while we have `DerefMut: Deref`)

We could use `Borrow` and `BorrowMut` with the small drawback to call `borrow()` every time (or use a function returning `&Transaction`). The other downside is that we can't provide an impl for deprecated `bip143::SigHashCache` like we do now.

What are the downsides of using `Deref`? (Apart from being discouraged to use by guidelines to emulate object oriented, which doesn't seem to be the case here)

Kixunil replied:
> `AsMut` is not a subtrait of `AsRef`

I don't see why that would be a problem. Just bound it by `AsRef<Transaction> + AsMut<Transaction>` where needed.

> What are the downsides of using `Deref`?

It's less general (e.g. `Box<PartiallySignedTransaction>` could theoretically implement both `AsRef<PartiallySignedTransaction>` and `AsRef<Transaction>` but it can never implement both `Deref`s), but honestly, I don't know how relevant it is. It's also a bit of stylistic issue - code in `std` doesn't use `Deref` in bounds.

I personally don't feel strongly about it.

RCasatta replied:
> I don't see why that would be a problem. Just bound it by AsRef<Transaction> + AsMut<Transaction> where needed.

Maybe you want to use `SigHashCache` having only the non-mutable borrow of `Transaction` and you don't need to use `access_witness`

RCasatta replied:
I agree `AsRef` + `AsMut` would be slightly better but it would make it worse to implement the deprecated method in `bip143.rs` that is using `Deref` and `DerefMut`. So I am leaving like it is unless someone feels strongly about it.

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+    outputs: sha256::Hash,
+}
+
+/// Values cached for segwit inputs, it's equal to [CommonCache] plus another round of `sha256`
```
nit: doc comments should use ``[`CommonCache`]`` style (the same applies to the above deprecated messages

RCasatta replied:
Using backtick for this and for other type references in the file in c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f

apoelstra replied:
Another nit: I think it's `CommonCache` that has the extra round of sha256, and `SegwitCache` has only one

apoelstra replied:
Oh, no, I am confusing segwit v1 with v0 aren't I. Never mind.

dr-orlovsky reviewed (commented):
Haven't finished full review in one go, just a questions for now

Kixunil commented on src/util/sighash.rs:
````diff
@@ -0,0 +1,904 @@
+    ///     sig_hasher.access_witness(inp).push(Vec::new());
+    /// }
+    /// ```
+    pub fn access_witness(&mut self, input_index: usize) -> &mut Vec<Vec<u8>> {
````
I believe the idiomatic name would be `witness_mut`

RCasatta replied:
Honestly here I kept the old signature type in bip143 module, we may change it to `witness_mut` and maybe also return `Option` similar to `Vec::get`.

@sanket1729 and @sgeisler do you think is worth changing this?

dr-orlovsky replied:
I think we need to change the signature to return optional (+change the name) - or introduce new function returning optional with old function provided only for legacy purposes with deprecated attr.

RCasatta replied:
Changed in `witness_mut` and returning an Option in f223be618fff7f06d6a1b2e0ca94c506f3ea2fd7

Kixunil commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+    ///     let _sighash = sig_hasher.signature_hash(inp, &prevout_script, 42, SigHashType::All);
+    ///     // ... sign the sighash
+    ///     sig_hasher.access_witness(inp).push(Vec::new());
+    /// }
```
Document that it panics if `input_index` is out of bounds?

RCasatta replied:
No need to comment anymore since now it returns an Option in f223be618fff7f06d6a1b2e0ca94c506f3ea2fd7 (commented the old method in `bip143.rs` though)

Kixunil commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+/// The `Annex` struct is a slice wrapper enforcing first byte to be `0x50`
+pub struct Annex<'a>(&'a [u8]);
```
`pub struct Annex([u8])` might be nicer but requires `unsafe` to use in practice.

RCasatta replied:
we have `#![forbid(unsafe_code)]` :)

Kixunil commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+            };
+            self.taproot_cache = Some(cache);
+        }
+        self.taproot_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
```
// ... and we produced `Some()` at the last line inside `if`.

RCasatta replied:
No need to change anymore since I included your PR https://github.com/RCasatta/rust-bitcoin/pull/3

RCasatta commented:
I am going to change doc-only stuff if another code change is required, otherwise, I'll do follow-up PR for the following:

* nit: doc comments should use [`CommonCache`] style (the same applies to the above deprecated messages
* // ... and we produced Some() at the last line inside if.
* Document that it panics if input_index is out of bounds? (if we are not switching to Option)

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+            };
+            self.taproot_cache = Some(cache);
+        }
+        self.taproot_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
```
I'd like to avoid unwraps when possible: for instance, somebody may change the logic in the lines above and this unwrap will suddenly start panicking.

So something like `match self.taproot_cache { None => { ... }, Some(tc) => tc }` will be much more preferable

Kixunil replied:
I was going to write this is impossible (I previously didn't write the exact comment you did because I believed so) but decided to double-check. I found [`Option::get_or_insert_with()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with), which is perfect for this case and I believe it should be used.

RCasatta replied:
The reasons I didn't use `get_or_insert_with` are here https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r670262033
If we agree on that, I should surely add a comment because it's already been suggested https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r671297559

Kixunil replied:
Heh, missed those. If my [suggested](https://github.com/RCasatta/rust-bitcoin/pull/3) solution to remove unwraps is not accepted then comments should definitely be written.

RCasatta replied:
Suggested solution accepted, thanks

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+        encode::consensus_encode_with_size(&self.0, writer)
+    }
+}
+
```
Are we sure that consensus-decoding Annex is not needed? They type seems to be usable outside of just signature creation scope

RCasatta replied:
I am not sure, maybe it's useful, however, I prefer to be added in another PR (implementing it would be non-breaking I think)

dr-orlovsky reviewed (approved):
utACK. Haven't checked the implementation match to BIPs though, will come to this later once we will have more ACKs and be sure that the code will not change.

Feel free to take my suggestions not for this PR but for the future follow-ups if there will be no need to change the code after other reviews.

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+//!
+//! Implementation of the algorithm to compute the message to be signed according to [Bip341](https://github.com/bitcoin/bips/blob/150ab6f5c3aca9da05fccc5b435e9667853407f4/bip-0341.mediawiki),
+//! [Bip143](https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki)
+//! and legacy
```
added `before Bip143` to clarify legacy meaning in c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+
+/// Hashtype of an input's signature, encoded in the last byte of the signature
+/// Fixed values so they can be casted as integer types for encoding
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
```
I am not sure on this, so I didn't change for now unless someone else will feel about it

RCasatta commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,904 @@
+
+impl SigHashType {
+    /// Break the sighash flag into the "real" sighash flag and the ANYONECANPAY boolean
+    pub(crate) fn split_anyonecanpay_flag(self) -> (SigHashType, bool) {
```
I don't think `split_anyonecanpay_flag` is that bad and is coherent with what we had before

RCasatta commented:
Sorry @dr-orlovsky , @sgeisler , @sanket1729  for another dismissed review but the change proposed by @Kixunil  is superior (especially since there were too many comments on why unwrapping before) and also the `access_witness` make sense for me (and potentially a breaking change to do it afterward).

Since I think the above two changes were important I fixed also documentations

sanket1729 reviewed (approved):
ACK c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f. Reviewed the diff from a37de1ade475e0c31c932121abaa7aec701b9987 which I previously ACKed

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+    }
+
+    /// Encode the BIP143 signing data for any flag type into a given object implementing a
+    /// [`std::io::Write`] trait.
```
Nit:

```suggestion
    /// [`io::Write`] trait.
```

as was specified above (to support no_std)

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+    }
+
+    /// Encode the legacy signing data for any flag type into a given object implementing a
+    /// [`std::io::Write`] trait. Internally calls [`Transaction::encode_signing_data_to`]
```
Ibid

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+    }
+
+    fn common_cache_minimal_borrow<'a>(
+        common_cache: &'a mut Option<CommonCache>,
+        tx: &R,
```
Why we are passing these two arguments when we can just use `self`?

Kixunil replied:
We can't because [`self.segwit_cache` may be borrowed](https://github.com/rust-bitcoin/rust-bitcoin/pull/628/files/c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f#diff-d38c556da0f54490a5799f38921a86b42a25dc5a26df0389d7a741ce95d33e58R587) during call to this function which prevents us to borrow `self` mutably. The name of the function is supposed to indicate it but maybe it's not clear enough? Have an idea for a better one?

dr-orlovsky replied:
Right, I see. Yes, I do not know any other workaround

dr-orlovsky reviewed (approved):
utACK c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f by diffing it to 6e06a32ccc70f13dc57c611cfcb6678ab520b9a0 having my ACK before.

Three nits: two doc related and one related to improvement of private method signature

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+            for txin in tx.input.iter() {
+                txin.previous_output
+                    .consensus_encode(&mut enc_prevouts)
+                    .unwrap();
```
Nit: I think it will be better to have an `expect` here

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+                txin.previous_output
+                    .consensus_encode(&mut enc_prevouts)
+                    .unwrap();
+                txin.sequence.consensus_encode(&mut enc_sequences).unwrap();
```
Ibid

dr-orlovsky commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+                outputs: {
+                    let mut enc = sha256::Hash::engine();
+                    for txout in tx.output.iter() {
+                        txout.consensus_encode(&mut enc).unwrap();
```
Ibid

dr-orlovsky commented:
@apoelstra we already have to ACKs here, but taking into account the importance of the code, probably it will be up to you to add the third ACK and merge the PR

JeremyRubin commented on src/util/bip143.rs:
```diff
@@ -241,11 +155,15 @@ impl<R: Deref<Target=Transaction>> SigHashCache<R> {
     /// regular reference, this method is available to allow modification to the witnesses.
     ///
     /// This allows in-line signing such as
+    ///
```
nit: comment belongs above line 163

JeremyRubin commented on src/util/bip143.rs:
````diff
@@ -263,13 +181,14 @@ impl<R: DerefMut<Target=Transaction>> SigHashCache<R> {
     /// ```
     pub fn access_witness(&mut self, input_index: usize) -> &mut Vec<Vec<u8>> {
-        &mut self.tx.input[input_index].witness
+        self.cache.witness_mut(input_index).unwrap()
````
nit: maybe you can just define an empty vec in the Self type that gets returned if input_index is out of bound. It's better than the unchecked access, and perhaps better than adding an unwrap / panic.

definitely a nit though, as this is deprecated code anyways.

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+    sequences: sha256::Hash,
+
+    /// in theory, `outputs` could be `Option` since `NONE` and `SINGLE` doesn't need it, but since
+    /// `ALL` is the mostly used variant by large, we don't bother
```
nit: the same comment is true for prevouts/sequences with ANYONECANPAY, no?

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+
+/// Values cached for segwit inputs, it's equal to [`CommonCache`] plus another round of `sha256`
+#[derive(Debug)]
+struct SegwitCache {
```
nit: SegwitV0Cache, taproot is a segwit too?

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+    /// `One` variant allows to provide the single Prevout needed. It's useful for example
+    /// when modifier `ANYONECANPAY` is provided, only prevout of the current input is needed.
+    /// The first `usize` argument is the input index this [`TxOut`] is referring to.
+    One(usize, &'u TxOut),
```
nit: in theory usize is redundant with &'u TxOut + Transaction as you can do the pointer math to compute where the TxOut lives in the out vector.

apoelstra replied:
I don't understand this, what vector are you talkng about?

JeremyRubin replied:
I guess i'm (incorrectly?) assuming that there is some backing `&'u [TxOut]` that this txout is resident from which the lifetime derives, (where ones we don't have are default() nulls), but i suppose that we might just have a single TxOut fetched in some contexts and not want a whole array, or that having a ref to the array would prevent some clever uses of splits and whatnot on that storage to hand out mutable and immutable references simultaneously...

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+    One(usize, &'u TxOut),
+    /// When `ANYONECANPAY` is not provided, or the caller is handy giving all prevouts so the same
+    /// variable can be used for multiple inputs.
+    All(&'u [TxOut]),
```
nit: a slice is the wrong type, we want a reference to the underlying storage to ensure we have the whole slice, not just partial.

apoelstra replied:
I strongly disagree

JeremyRubin replied:
Another way to make this API safe would be via a smart constructor. E.g.

```rust
pub enum Prevouts<'u> {
//...
All(SafeSliceAll<'u>)
}
use private::SafeSliceAll;
mod private {
// self.0 not pub so can only be constructed locally.
pub struct SafeSliceAll<'u>(&'u [TxOut]);
impl<'u> From<&'u Vec<TxOut>> for SafeSliceAll<'u> {
//....
}
}
```

Allowing to construct an All from a arbitrary slice is error prone IMO since you can't guarantee it's *actually* all.

JeremyRubin replied:
we need to use such a smart constructor type because pub enums require all fields of variants are pub iirc.

apoelstra replied:
I don't understand why the user would be required to have the txouts all in a `Vec` like this.

JeremyRubin commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,909 @@
+
+impl<'u> Prevouts<'u> {
+    fn check_all(&self, tx: &Transaction) -> Result<(), Error> {
+        if let Prevouts::All(prevouts) = self {
```
nit: as noted earlier, a non-slice type would eliminate this check.

apoelstra replied:
This isn't a check, it's a destructuring.

JeremyRubin replied:
Ah the check is the line below

JeremyRubin commented:
CR-ack; some nits nothing needing fixing IMO

apoelstra commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+    /// 0x3: Sign the output whose index matches this input's index. If none exists,
+    /// sign the hash `0000000000000000000000000000000000000000000000000000000000000001`.
+    /// (This rule is probably an unintentional C++ism, but it's consensus so we have
+    /// to follow it.)
```
Nit: we should update ths comment to say that this is only true for legacy input. For segwit (v0 and taproot), if you try to SINGLE-sign an output that doesn't exist it'll just fail.

apoelstra commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+        //      sha_single_output (32): the SHA256 of the corresponding output in CTxOut format.
+        if sighash == SigHashType::Single {
+            let mut enc = sha256::Hash::engine();
+            self.tx.output[input_index].consensus_encode(&mut enc)?;
```
Need to bounds-check this, either here or in `check_index` above.

apoelstra replied:
Ah this is fixed in a later commit.

apoelstra commented on src/util/sighash.rs:
```diff
@@ -0,0 +1,836 @@
+        } else if sighash == LegacySigHashType::Single && input_index < self.tx.output.len() {
+            let mut single_enc = SigHash::engine();
+            self.tx.output[input_index].consensus_encode(&mut single_enc)?;
+            SigHash::from_engine(single_enc).consensus_encode(&mut writer)?;
```
Should add an `else if sighash == LegacySigHashType::Single && input_index >= self.tx.output.len()` which encodes the one hash

apoelstra replied:
Never mind.

apoelstra reviewed (approved):
ACK c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f

This is really great!

Not going to hold up merging for nits about comments etc
